content/blog/2020/05/2020-05-07-docker-windows-agents.adoc

```diff
@@ -0,0 +1,116 @@
+
+```
+java -jar agent.jar -jnlpUrl https://JENKINS_URL/jenkins/computer/AGENTNAME/slave-agent.jnlp -secret SOMESECRETVALUE -workDir "C:\Users\jenkins\Agent"
+```
+
+This can be passed to the `docker run` command to connect the agent to the master (make sure to specify the correct values for JENKINS_URL, SOMESECRETVALUE, AGENTNAME and the workDir, which should match the value you specified in agent setup above).
+
+```
+docker run --name agent jenkins/agent:jdk8-nanoserver-1809 java -jar agent.jar -jnlpUrl https://JENKINS_URL/jenkins/computer/AGENTNAME/slave-agent.jnlp -secret SOMESECRETVALUE -workDir "C:\Users\jenkins\Agent"
+```
+
+You can also specify a volume for the work directory if desired (e.g., to keep certain files around in cached form or similar).
+
+```
+docker run -i --rm --name agent -v agent-workdir:C:/Users/jenkins/Work jenkins/agent java -jar C:/ProgramData/Jenkins/agent.jar -jnlpUrl https://JENKINS_URL/jenkins/computer/AGENTNAME/slave-agent.jnlp -secret SOMESECRETVALUE -workDir C:/Users/jenkins/Work
+```
+
+Just make sure this directory doesn't already exist or is empty inside the container (this is a requirement of volumes with Docker on Windows).
+
+
+
+=== jenkins/inbound-agent
+
+The inbound-agent Docker image tries to provide a higher level interaction with the agent.jar executable. It provides a PowerShell wrapper script around agent.jar and it is specified as the entrypoint so that you just need to pass in some command line arguments to run the agent. Several of these can also be passed using environment variables instead (they are listed in all capitals).
+
+Command line arguments include:
+
+* `-Url <URL>` or JENKINS_URL - the Jenkins master URL (this should NOT include the computer/NAME/slave-agent.jnlp part)
+* `-Secret <SECRET>` or JENKINS_SECRET - the secret as shown on the master after creating the agent
+* `-Name <NAME>` or JENKINS_AGENT_NAME - the name of the agent, it should match the name you specified when creating the agent on the master
+* `-Tunnel <HOST:PORT>` or JENKINS_TUNNEL - A tunnel host to route the TCP traffic through when connecting to the master
+* `-WorkDir <WORKDIR>` or JENKINS_AGENT_WORKDIR - same as the -workDir parameter mentioned above in the jenkins/agent image information.
+* `-WebSocket` or JENKINS_WEB_SOCKET=true - when present the connection to the master will be done via WebSocket instead of TCP networking.
+* `-DirectConnection <HOST:PORT>` or JENKINS_DIRECT_CONNECTION - Connect directly to this TCP agent port, skipping the HTTP(S) connection parameter download.
+* `-InstanceIdentity <BASE64VALUE>` or JENKINS_INSTANCE_IDENTITY - The base64 encoded InstanceIdentity byte array of the Jenkins master. When this is set, the agent skips connecting to an HTTP(S) port for connection info
+* `-Protocols <PROTOCOLS>` or JENKINS_PROTOCOLS - Specify the remoting protocols to attempt when instanceIdentity is provided
```

Comment:
What are the allowed values for protocols?

I assume some way of saying "WebSocket" is one of them.  Is there also some way of saying "Jenkins Remoting"?

Reply:
Good question, it looks like the Windows agents don't currently actually support that parameter (would be an easy fix to do it). It's an option passed to the agent.jar, I will find out what is supported there and also create a PR to actually use the `-Protocols` value

Reply:
Added link to the list of protocols in the documentation from the remoting repository and a note about the issue with the Protocols parameter.